Documentation and feature flag

jens-esr opened the pull request:
Added the ability to switch between thread local and a split socket via a feature flag.

An assert was added in the builder if only one of the two flags is set. 
Is there a way to check this at compile time?

Documentation was added. 

Currently there is still a compiler warning that UnixDatagram is not used when compiling for a shared socket.

jens-esr committed:
Increase buffer size for null byte

jens-esr committed:
Change copy of tag to reference

jens-esr committed:
Change socket to private

jens-esr committed:
Reformat code

jens-esr committed:
Add documentation for build

jens-esr committed:
Add documentation for builder

jens-esr committed:
Add documentation for Buffer

jens-esr committed:
Add module description

jens-esr committed:
Fix typo

jens-esr committed:
Add static socket

jens-esr committed:
Remove use statement from doc

jens-esr committed:
Change format

jens-esr committed:
Use static locked socket

jens-esr committed:
Remove copy of string

jens-esr committed:
Remove use from doc

jens-esr committed:
Add features

jens-esr committed:
Add assert to builder

jens-esr committed:
Add readme

jens-esr committed:
Add different OS behavior to readme

flxo commented on README.md:
```diff
@@ -1,4 +1,67 @@
+Logd-logger is configured once when initiating.
+Messages are sent using the macros `error!`, `warn!`, `info!`, `debug!` and `trace!`.
+
+## Android
+In case logd-logger is compiled for Android, 
+the log messages are sent to Android's logd socket.
+
+## Other OS
+For other Operating Systems the logging messages are sent to standard out.
+These are displayed in logd format.
+
+# Usage
+
```
Inconsistent newlines after the section.

flxo commented on Cargo.toml:
```diff
@@ -18,6 +18,7 @@ keywords = [
 env_logger = { version = "0.8.2", features = ["regex"], default-features = false }
 log = { version = "0.4.11", features = ["std"] }
 time = "0.2.23"
+lazy_static = { version = "1.4.0", optional = true }
```
The deps should be in alphanumeric order.

flxo commented on Cargo.toml:
```diff
@@ -30,3 +31,8 @@ winapi = { version = "0.3.9", features = ["processthreadsapi"] }
+[features]
+default = ["locked"]
+tls = []
+locked = ["lazy_static"]
```
`locked` is probably not the right name. With this  feature one socket is shared. Rename this feature to `shared`.

flxo commented on README.md:
```diff
@@ -1,4 +1,67 @@
+
+## Other OS
+For other Operating Systems the logging messages are sent to standard out.
+These are displayed in logd format.
```
I'd replace `logd format` with `logcat` and add a link.

flxo commented on Cargo.toml:
```diff
@@ -18,6 +18,7 @@ keywords = [
+lazy_static = { version = "1.4.0", optional = true }
 
 [target.'cfg(target_os = "android")'.dependencies]
 bytes = "1.0.0"
```
We should change this to `bytes = "1"`

flxo commented on README.md:
```diff
@@ -1,4 +1,67 @@
+Next:
+In the example the logging is initialized. First a [`Builder`] is created.
+Then the log level is set. All messages with lower log level are dropped.
+Then a tag is set which is added in front of each log message.
```
This is grammatically wrong ;-)

jens-esr replied:
I don't get what is wrong.

flxo commented on README.md:
```diff
@@ -1,4 +1,67 @@
+Next, the module name is set to be prepended.
+After the configuration, it is tried to initialize the global logger.
+
+Afterwards five different macros can be used for logging.
```
This can be omitted since everyone who uses this crates know the Rust logger concept.

flxo commented on src/lib.rs:
```diff
@@ -1,10 +1,93 @@
+#[cfg(all(feature = "locked", not(feature = "tls"), target_os = "android"))]
+lazy_static! {
+    static ref SOCKET: std::os::unix::net::UnixDatagram = {
+        let socket = std::os::unix::net::UnixDatagram::unbound().expect("Failed to create socket");
```
Add a `use std::os::unix::net::UnixDatagram`. It's used twice and reads bulky.

flxo commented on src/lib.rs:
```diff
@@ -1,10 +1,93 @@
+    static ref SOCKET: std::os::unix::net::UnixDatagram = {
+        let socket = std::os::unix::net::UnixDatagram::unbound().expect("Failed to create socket");
+        socket
+            .connect("/dev/socket/logdw")
```
Refactor the path to the socket into a constant.

flxo commented on src/lib.rs:
```diff
@@ -71,16 +154,23 @@ impl From<log::Level> for Priority {
 }
 
-/// Log buffer ids
+/// Different Log Buffer for Android
```
It's still `Log buffer ids` This is the term that is used within Android.

flxo commented on src/lib.rs:
```diff
@@ -379,16 +464,24 @@ impl Log for Logger {
 
-            SOCKET.with(|f| f.send(&buffer).expect("Logd socket error"));
+            #[cfg(all(feature = "tls", not(feature = "locked"), target_os = "android"))]
+            {
```
Think this bracket is not needed.

flxo commented on src/lib.rs:
```diff
@@ -379,16 +464,24 @@ impl Log for Logger {
+            }
+
+            #[cfg(all(feature = "locked", not(feature = "tls"), target_os = "android"))]
+            {
```
Think this bracket is not needed.

flxo commented on src/lib.rs:
```diff
@@ -1,10 +1,93 @@
+//! `logd-logger`
```
Theres a crate that generates the Readme out of `lib.rs`. Can we use that in order to not maintain the same doc twice?

jens-esr replied:
There is a cargo plugin that does the trick. Should we use that? [cargo-sync-readme](https://crates.io/crates/cargo-sync-readme)

flxo replied:
Sounds like a plan. Go for it!

jens-esr committed:
Change feature locked to shared

jens-esr committed:
Remove newline

jens-esr committed:
Change dependency order

jens-esr committed:
Add link to logcat

jens-esr committed:
Change version of bytes dependency

jens-esr committed:
Add missing dependencies

jens-esr committed:
Let compiler resolve type

jens-esr committed:
Move logdw path to constant

jens-esr committed:
Change buffer documentation

jens-esr committed:
Remove unnecessary bracket

jens-esr committed:
Remove unnecessary documentation

jens-esr committed:
Add synchronization between readme and doc

Use the cargo plugin sync-readme for synchronization between rustdoc and Readme.md
